File: scripts/build_offline_db.py
#!/usr/bin/env python3
import json
import os
from youtube_transcript_api import YouTubeTranscriptApi
import time

METADATA_FILE = "src/data/offline_se_videos_metadata.json"
OUTPUT_FILE = "src/data/offline_transcripts.json"

def fetch_and_translate(video_id):
    try:
        api = YouTubeTranscriptApi()
        tlist = api.list(video_id)
        
        # Tìm tiếng Anh gốc
        try:
            source_t = tlist.find_transcript(['en', 'en-US', 'en-GB'])
        except:
            source_t = list(tlist)[0]

        en_cues = source_t.fetch()
        
        # Dịch song ngữ bằng AI nội bộ YouTube
        try:
            vi_cues = source_t.translate('vi').fetch()
        except:
            vi_cues = en_cues

        cues_list = []
        for idx in range(len(en_cues)):
            en_cue = en_cues[idx]
            vi_cue = vi_cues[idx] if idx < len(vi_cues) else en_cue

            text_en = en_cue['text'].replace("\n", " ").strip()
            if not text_en:
                continue

            cues_list.append({
                "id": idx + 1,
                "start": round(float(en_cue['start']), 2),
                "duration": round(float(en_cue['duration']), 2),
                "end": round(float(en_cue['start']) + float(en_cue['duration']), 2),
                "textEn": text_en,
                "textVi": vi_cue['text'].replace("\n", " ").strip(),
                "words": [w for w in text_en.split() if w]
            })
        return cues_list
    except Exception as e:
        print(f"Lỗi tải video {video_id}: {str(e)}")
        return []

def build_db():
    print("🚀 Bắt đầu tiến trình tải dữ liệu Offline...")
    with open(METADATA_FILE, 'r', encoding='utf-8') as f:
        videos = json.load(f)

    # Đọc db cũ nếu có
    db = {}
    if os.path.exists(OUTPUT_FILE):
        with open(OUTPUT_FILE, 'r', encoding='utf-8') as f:
            try:
                db = json.load(f)
            except:
                pass

    for i, v in enumerate(videos):
        vid = v['videoId']
        if vid in db and len(db[vid]) > 0:
            print(f"[{i+1}/{len(videos)}] Đã có sẵn: {vid} - Bỏ qua.")
            continue
            
        print(f"[{i+1}/{len(videos)}] Đang cào phụ đề cho: {vid} - {v['title']}...")
        cues = fetch_and_translate(vid)
        if cues:
            db[vid] = cues
            print(f"   -> Thành công: Tải được {len(cues)} câu song ngữ.")
        else:
            print(f"   -> Thất bại: Không lấy được phụ đề.")
            
        time.sleep(1) # Tránh bị YouTube block do spam

    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(db, f, ensure_ascii=False, indent=2)
        
    print(f"✅ Hoàn tất! Đã đóng gói thành công tổng cộng {len(db.keys())} bộ phụ đề vào {OUTPUT_FILE}")

if __name__ == "__main__":
    build_db()

File: scripts/build_offline_db.mjs
import fs from 'fs'
import path from 'path'

const METADATA_FILE = path.join(process.cwd(), 'src/data/offline_se_videos_metadata.json')
const OUTPUT_FILE = path.join(process.cwd(), 'src/data/offline_transcripts.json')

const INSTANCES = [
  'https://invidious.nerdvpn.de',
  'https://vid.puffyan.us',
  'https://invidious.drgns.space',
  'https://invidious.adminforge.de',
  'https://invidious.slipfox.xyz',
  'https://invidious.protokolla.fi',
  'https://iv.ggtyler.dev',
]

async function fetchFromInvidious(videoId) {
  for (const instance of INSTANCES) {
    try {
      const controller = new AbortController()
      const timeout = setTimeout(() => controller.abort(), 8000)
      const res = await fetch(`${instance}/api/v1/captions/${videoId}?format=vtt`, { signal: controller.signal })
      clearTimeout(timeout)
      
      if (!res.ok) continue
      const data = await res.json()
      
      // Lấy phụ đề tiếng Anh
      const enCaption = data.captions?.find(c => c.language_code.startsWith('en'))
      if (!enCaption) continue
      
      // Tải nội dung
      const vttRes = await fetch(instance + enCaption.url)
      if (!vttRes.ok) continue
      
      const text = await vttRes.text()
      if (text.includes('WEBVTT')) {
        return text
      }
    } catch (e) {
      continue
    }
  }
  return null
}

function parseVttToCues(vttText) {
  const lines = vttText.split('\n')
  const cues = []
  let currentCue = null
  let idCounter = 1

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim()
    if (!line || line === 'WEBVTT') continue

    if (line.includes('-->')) {
      const parts = line.split('-->')
      const start = parseTime(parts[0])
      const end = parseTime(parts[1])
      
      currentCue = {
        id: idCounter++,
        start,
        end,
        duration: end - start,
        textEn: '',
        textVi: '... (Bản dịch đang được tải) ...', // Offline data, tạm thời
        words: []
      }
      cues.push(currentCue)
    } else if (currentCue) {
      if (currentCue.textEn) currentCue.textEn += ' ' + line
      else currentCue.textEn = line
      
      // Cleanup tags like <c> or </c>
      currentCue.textEn = currentCue.textEn.replace(/<[^>]+>/g, '').trim()
    }
  }
  
  // Format words
  for (const c of cues) {
    c.words = c.textEn.split(/\s+/).filter(Boolean)
  }

  return cues
}

function parseTime(timeStr) {
  const t = timeStr.trim().split(':')
  if (t.length === 3) {
    return parseFloat(t[0]) * 3600 + parseFloat(t[1]) * 60 + parseFloat(t[2])
  } else if (t.length === 2) {
    return parseFloat(t[0]) * 60 + parseFloat(t[1])
  }
  return 0
}

async function run() {
  console.log('🚀 Bắt đầu cào dữ liệu Offline bằng Invidious...')
  const rawMetadata = fs.readFileSync(METADATA_FILE, 'utf-8')
  const videos = JSON.parse(rawMetadata)
  
  let db = {}
  if (fs.existsSync(OUTPUT_FILE)) {
    db = JSON.parse(fs.readFileSync(OUTPUT_FILE, 'utf-8'))
  }

  for (let i = 0; i < videos.length; i++) {
    const vid = videos[i].videoId
    if (db[vid] && db[vid].length > 0) {
      console.log(`[${i + 1}/${videos.length}] Đã có sẵn: ${vid} - Bỏ qua.`)
      continue
    }

    console.log(`[${i + 1}/${videos.length}] Đang cào phụ đề cho: ${vid} - ${videos[i].title}...`)
    const vtt = await fetchFromInvidious(vid)
    if (vtt) {
      const cues = parseVttToCues(vtt)
      if (cues.length > 0) {
        db[vid] = cues
        console.log(`   -> Thành công: Tải được ${cues.length} câu.`)
        // Save incremental
        fs.writeFileSync(OUTPUT_FILE, JSON.stringify(db, null, 2))
      } else {
        console.log(`   -> Lỗi: Phân tích VTT rỗng.`)
      }
    } else {
      console.log(`   -> Thất bại: Invidious API không trả về dữ liệu.`)
    }
    
    await new Promise(r => setTimeout(r, 1000))
  }
  
  console.log(`✅ Hoàn tất! Đã lưu toàn bộ vào ${OUTPUT_FILE}`)
}

run()
